File: Homework-1/an_nguyen_task1.scala
import java.io.{File, PrintWriter}
import org.json4s._
import org.json4s.jackson.JsonMethods._
import org.apache.log4j.{Level, Logger}
import org.apache.spark.sql.SparkSession


object an_nguyen_task1 {
	implicit val formats: DefaultFormats.type = DefaultFormats

	case class User(user_id: String, name: String, review_count: Int, yelping_since: String, friends: Any,
					useful: Int, funny: Int, cool: Int, fans: Int, elite: Any, average_stars: Float,
					compliment_hot: Int, compliment_more: Int, compliment_profile: Int, compliment_cute: Int,
					compliment_list: Int, compliment_note: Int, compliment_plain: Int, compliment_cool: Int,
					compliment_funny: Int, compliment_writer: Int, compliment_photos: Int
				   )

	def main(args: Array[String]): Unit = {
		val pw = new PrintWriter(new File(args(1)))
		pw.write("{")

		Logger.getLogger("org").setLevel(Level.INFO)
		val spark = SparkSession
			.builder()
			.appName("an_nguyen_task1")
			.config("spark.master", "local[*]")
			.getOrCreate()

		val data = spark
			.sparkContext
			.textFile(args(0))
			.coalesce(8)
			.map(convertToJson)
			.map(x => (x.user_id, x.name, x.review_count, x.yelping_since))

		val total_users: Long = data.count()
		pw.write("\"total_users\": " + total_users + ",")

		val total_reviews = data.map(x => x._3).reduce((a, b) => (a + b))
		val avg_reviews = total_reviews.toString.toFloat / total_users.toFloat
		pw.write("\"avg_reviews\": " + avg_reviews + ",")

		val names = data.map(x => x._2).groupBy(x => x)
		pw.write("\"distinct_usernames\": " + names.count() + ",")

		val num_users = data.map(x => x._4).filter(x => x.contains("2011")).count()
		pw.write("\"num_users\": " + num_users + ",")

		pw.write("\"top10_popular_names\": [")
		val top_names = names.mapValues(x => x.size).sortBy(_._2, ascending = false).take(10)
		for (i <- top_names) {
			pw.write("[\"" + i._1 + "\", " + i._2 + "]")
			if (top_names.indexOf(i) < top_names.length - 1) {
				pw.write(",")
			}
		}
		pw.write("],")

		pw.write("\"top10_most_reviews\": [")
		val top_reviewers = data.map(x => (x._1, x._3)).sortBy(x => (-x._2, x._1)).take(10)
		for (i <- top_reviewers) {
			pw.write("[\"" + i._1 + "\", " + i._2 + "]")
			if (top_reviewers.indexOf(i) < top_reviewers.length - 1) {
				pw.write(",")
			}
		}
		pw.write("]")

		pw.write("}")
		pw.close()
	}

	def convertToJson(line: Any): User = {
		val temp = line + ""
		parse(temp).extract[User]
	}
}

File: Homework-1/an_nguyen_task1.py
import sys
import json
from pyspark.sql import SparkSession


def main(input_file, output_file):
    result = {}

    spark = SparkSession.builder.master("local[*]").appName("HW1").getOrCreate()
    text = spark                                \
        .sparkContext                           \
        .textFile(input_file)                   \
        .coalesce(8)                            \
        .map(json.loads)                        \
        .map(lambda x: [x["user_id"],
                        x["name"],
                        x["review_count"],
                        x["yelping_since"]])    \
        .cache()

    # A. Find the total number of users
    result["total_users"] = text.count()

    # B. Find the average number of written reviews of all users
    total_reviews = text.map(lambda x: x[2]).reduce(lambda a, b: a + b)
    result["avg_reviews"] = total_reviews / result["total_users"]

    # C. Find the number of distinct user names
    names = text.map(lambda x: x[1]).groupBy(lambda x: x).cache()
    result["distinct_usernames"] = names.count()

    # D. Find the number of users that joined yelp in the year 2011
    result["num_users"] = text.map(lambda x: x[3]).filter(lambda x: "2011" in x).count()

    # E. Find Top 10 popular names and the number of times they appear
    result["top10_popular_names"] = names.mapValues(len).sortBy(lambda x: (-x[1], x[0])).take(10)

    # F. Find Top 10 user ids who have written the most number of reviews
    result["top10_most_reviews"] = text.map(lambda x: [x[0], x[2]]).sortBy(lambda x: (-x[1], x[0])).take(10)

    with open(output_file, "w") as f:
        json.dump(result, f)


if __name__ == "__main__":
    main(sys.argv[1], sys.argv[2])
